src/class_loader_core.cpp

```diff
@@ -57,8 +57,26 @@ std::recursive_mutex & getPluginBaseToFactoryMapMapMutex()
 
 BaseToFactoryMapMap & getGlobalPluginBaseToFactoryMapMap()
 {
-  static BaseToFactoryMapMap instance;
-  return instance;
+
+  static std::shared_ptr<BaseToFactoryMapMap> instance;
+
+  // TODO: need lock?
+  // TODO: we may use unique_ptr but use shared_ptr for PoC
```

Comment:
@y-okumura-isp `std::unique_ptr` sounds best. Also, the custom deleter is fine, using `std::unique_ptr<MetaObject>` values would be best.

Reply:
Thank  you for comment.
I intended to make changes minimum. But it looks better to hold smart pointer directory.

I have one confirmation.

> std::unique_ptr<MetaObject> values would be best.

What you say is to change

```
typedef std::map<ClassName, impl::AbstractMetaObjectBase *> FactoryMap;
```

to

```
typedef std::map<ClassName, std::unique_ptr<impl::AbstractMetaObjectBase>> FactoryMap;
```

Right?

The pointer `FactoryMap[class]` is used in `createInstance`, so we could `shared_ptr`.

```
typedef std::map<ClassName, std::shared_ptr<impl::AbstractMetaObjectBase>> FactoryMap;
```

Reply:
> Right?

Correct.

> The pointer FactoryMap[class] is used in createInstance, so we could shared_ptr.

I presume you mean https://github.com/ros/class_loader/blob/8864be372a83921f15906ae80752be9d6a300b81/include/class_loader/class_loader_core.hpp#L251. If so, I see no need for shared ownership. Just grab the raw pointer momentarily.